### PR TITLE
Fixes #39241 - Add Katello 4.21 upgrade task to convert Python-type repositories for publication-less Pulp

### PR DESCRIPTION
**Pull request by qcjames53**
#### What are the changes introduced in this pull request?
Adds a rake upgrade task to:

1. Migrate Python-type repo distribution from publication to repository_version.
2. Wait for task to complete.
3. Clear publication_href from Katello DB.
4. Delete Python publications from Pulp DB (needed for pulp_python routing reasons).

#### Considerations taken when implementing this change?

- This job does not clear orphaned publications, which may be an issue. The best way to do that is full orphan cleanup, which I suggest should be ran at the same time as this. Katello upgrades should take care of Pulp orphans, I think.

#### What are the testing steps for this pull request?

1. Upgrade to Pulp 3.105. This should be by default on new katello-devel boxes these days.
5. Create a Python repo complete with publication in Pulp's db. Claude generated me a biblically-accurate shell script that I can share if you'd like. Be not afraid.
6. Run the rake task with `bundle exec rake katello:upgrades:4.21:cleanup_python_publications`. Check for rake task success. Ensure the python repo is clean, the Pulp DB is clean, and the publication was deleted. If there are ANY remaining publications, pulp will have issues after migration. Orphan cleanup could be required depending on the pre-testing setup but let me know if that's the case since there's a chance it's an issue.
7. Run `ktest katello/test/lib/tasks/upgrades/4.21/cleanup_python_publications_test.rb` and make sure it passes.
8. Run an upgrade from 4.20 to 4.21 to make sure this task gets called (full disclosure I have not done this yet).
9. Vet the source and unit tests for quality.

## Summary by CodeRabbit

* **New Features**
  * Added an upgrade task for v4.21 to clean up Python publications and migrate distributions as part of the upgrade.

* **Tests**
  * Added a comprehensive test suite covering publication listing/deletion, distribution migration polling, error handling, and edge cases (missing proxy, empty lists, not-found responses).

### CODE REVIEW

**Comment by qcjames53** [comment]
The JIRA says we should regenerate metadata for Python repos in Katello but I found it a lot easier to just clear out all the publication_hrefs. Is there a reason we were advising the former approach that I'm not considering?

**Comment by sjha4** [comment]
I am a little confused by the flow on this. If we delete all publications on upgrade, we'll have distributions which are not yet pointed to repo versions. We would want to call refresh distribution on python distributions to migrate them to repo versions before we delete publications in Katello and in pulp. Does that make sense?

**Comment by sjha4** [comment]
Some merge conflicts following the merge of the other PR..

**Comment by qcjames53** [comment]
I'll take care of this after lunch :+1:

**Comment by qcjames53** [comment]
I've updated the description + testing steps to accurately reflect the PR changes. Please let me know if the console output is too wordy during the upgrade/rake job.

**Comment by qcjames53** [comment]
I'm happy with the current state of the PR. This is good to test :)

File: lib/katello/tasks/upgrades/4.21/cleanup_python_publications.rake [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                puts "Task #{index + 1}/#{tasks.count} completed successfully"
+                break
+              elsif task_status.state == 'failed'
+                puts "WARNING: Task #{index + 1}/#{tasks.count} failed: #{task_status.error}"
+                break
+              end
+
+              puts "."
+              sleep 1
+            end
```
[the remark this answers is not in the document]
Comment (qcjames53) [reply lib/katello/tasks/upgrades/4.21/cleanup_python_publications.rake]: @sjha4 What are your thoughts on this? Would you rather hang the upgrade or throw an error?

**Comment by sjha4** [comment]
Ran some tests with python data..Got claude to run some pre and post verifications and claude looks happy.. Will run some more tests and try some capsule syncs and see if I see anything noteworthy..
 ```
 Final Verification Summary                                                                                                                      
                  
  ✅ All 22 distributions use repository_version (not publication)                                                                                
  ✅ All 32 Katello repos have publication_href: nil
  ✅ All 32 Katello repos point to correct version_href                                                                                           
  ✅ All base_paths map correctly to Katello repository expectations                                                                              
  ✅ All PyPI URLs serve content correctly                                                                                                        
  ✅ All 10 publications deleted from Pulp                                                                                                        
                                                                                                                                                  
  Distribution sharing breakdown:                                                                                                                 
  - 5 base_paths with 2+ Katello repos sharing: cv1/Library (4×), cv1/Library requests (3×), cv1/Library color (4×), cv2/Library (2× each)
  - 17 base_paths with 1 Katello repo each                                                                                                        
                                          
  The migration is 100% correct! 🎉
```

File: lib/katello/tasks/upgrades/4.21/cleanup_python_publications.rake (hunk 1 above)
[the remark this answers is not in the document]
Comment (sjha4) [reply lib/katello/tasks/upgrades/4.21/cleanup_python_publications.rake]: Throwing an error seems better than hanging forever..On that note, reusing our pulp task service might simplify this task polling..Something like this:
```
  task_objects = tasks.map { |task_response| Katello::Pulp3::Task.new(smart_proxy, task_response) }                                               
                                                                                                   
  task_objects.each_with_index do |task, index|                                                                                                   
    until task.done?                           
      print "."                                                                                                                                   
      sleep 1     
      task.poll                                                                                                                                   
    end        
                                                                                                                                                  
    if task.error 
      puts "\nWARNING: Task #{index + 1}/#{tasks.count} failed: #{task.error}"
    else                                                                      
      puts "\nTask #{index + 1}/#{tasks.count} completed successfully"                                                                            
    end                                                               
  end                            
```

File: lib/katello/tasks/upgrades/4.21/cleanup_python_publications.rake [hunk 2]
```diff
@@ -0,0 +1,127 @@
+namespace :katello do
+  namespace :upgrades do
+    namespace '4.21' do
+      desc "Destroy all PythonPublication objects from Pulp database"
+      task :cleanup_python_publications => ["environment"] do
+        User.current = User.anonymous_api_admin
+
+        python_repos = Katello::Repository.joins(:root).where(katello_root_repositories: { content_type: 'python' })
```
Comment (sjha4) [inline lib/katello/tasks/upgrades/4.21/cleanup_python_publications.rake:8]: Can we exit early if there are no python repos? We can assume orphan cleanup takes care of any deleted repos if there was python content in the past and cases like that..

Comment (qcjames53) [reply lib/katello/tasks/upgrades/4.21/cleanup_python_publications.rake:8]: I think that's fair!

**Comment by qcjames53** [comment]
This newest push should address all of your concerns!

File: lib/katello/tasks/upgrades/4.21/cleanup_python_publications.rake [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        if python_repos.empty?
+          puts "No python repos found. Skipping upgrade."
```
Comment (sjha4) [inline lib/katello/tasks/upgrades/4.21/cleanup_python_publications.rake]:
```suggestion
          puts "No python repos found. Skipping upgrade task."
```

Comment (qcjames53) [reply lib/katello/tasks/upgrades/4.21/cleanup_python_publications.rake]: Good catch :grin: . Fixed!

File: lib/katello/tasks/upgrades/4.21/cleanup_python_publications.rake [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+        smart_proxy = SmartProxy.pulp_primary
+        unless smart_proxy
+          puts "ERROR: Primary Pulp smart proxy not found."
+          exit 1
+        end
+
+        migrate_distributions(python_repos, smart_proxy)
+        clear_publication_hrefs(python_repos)
+        delete_python_publications(smart_proxy)
```
Comment (sjha4) [inline lib/katello/tasks/upgrades/4.21/cleanup_python_publications.rake:21]: The coderabbit comment about this order makes sense..Let's switch this to delete_python_publications before clearing katello DB references to them with clear_publication_hrefs.

Comment (qcjames53) [reply lib/katello/tasks/upgrades/4.21/cleanup_python_publications.rake:21]: That makes sense to me. I've swapped the two operations. I also implemented the other coderabbit suggestion of failing the upgrade if python distributions cannot be migrated to repository versions from publications. Unit tests are updated to reflect this.

**Review by sjha4 — APPROVED** [review]
Task successfully migrates repos to use distros with repo versions and cleans up all publications and references. Ack. ✔️
